chef/cookbooks/nova/recipes/ceph.rb

```diff
@@ -0,0 +1,187 @@
+        secret_list = virsh_secret.run_command.stdout
+        virsh_secret.error!
+
+        secret_lines = secret_list.strip.split("\n")
+        if secret_lines.length < 2 || !secret_lines[0].start_with?("UUID") || !secret_lines[1].start_with?("----")
+          raise "cannot fetch list of libvirt secret"
+        end
+        secret_lines.shift(2)
+
+        secret_lines.each do |secret_line|
+          secret_uuid = secret_line.split(" ")[0]
+          cmd = ["virsh", "secret-dumpxml", secret_uuid]
+          virsh_secret_dumpxml = Mixlib::ShellOut.new(cmd)
+          secret_xml = virsh_secret_dumpxml.run_command.stdout
+          # some secrets might not be ceph-related, skip these
+          next if secret_xml.index("<usage type='ceph'>").nil?
+
+          # lazy xml parsing
+          re_match = %r[<usage type='ceph'>.*<name>(.*)</name>]m.match(secret_xml)
+          next if re_match.nil?
+          secret_usage = re_match[1]
+          undefine = false
+
+          if secret_uuid == rbd_uuid
+            undefine = true if secret_usage != "client.#{rbd_user} secret"
+          else
+            undefine = true if secret_usage == "client.#{rbd_user} secret"
+          end
+
+          if undefine
+            cmd = ["virsh", "secret-undefine", secret_uuid]
+            virsh_secret_undefine = Mixlib::ShellOut.new(cmd)
+            virsh_secret_undefine.run_command
+          end
+        end
+
+        if !admin_keyring.empty? && File.exists?(admin_keyring)
+          # Now add our secret and its value
+          cmd = ["ceph", "-k", admin_keyring, "-c", ceph_conf, "auth", "get-key", "client.#{rbd_user}" ]
+          ceph_get_key = Mixlib::ShellOut.new(cmd)
+          client_key = ceph_get_key.run_command.stdout.strip
+          ceph_get_key.error!
+        else
+          # Check if rbd keyring was uploaded manually by user
+          client_keyring = "/etc/ceph/ceph.client.#{rbd_user}.keyring"
+          if File.exists?(client_keyring)
+            f = File.open(client_keyring)
+            f.each do |line|
+              if match = line.match("key\s*=\s*(.+)")
+                client_key = match[1]
+                break
+              end
+            end
+          end
+        end
+
+        cmd = ["virsh", "secret-get-value", rbd_uuid ]
+        virsh_secret_get_value = Mixlib::ShellOut.new(cmd)
+        secret = virsh_secret_get_value.run_command.stdout.chomp.strip
+
+        if secret != client_key
+          secret_file_path = "/etc/ceph/ceph-secret-#{rbd_uuid}.xml"
+          secret_xml_file =  Chef::Resource::File::File.new(secret_file_path, run_context)
+          secret_xml_file.owner  "root"
+          secret_xml_file.group  "root"
+          secret_xml_file.mode   "0640"
+          secret_xml_file.content "<secret ephemeral='no' private='no'> <uuid>#{rbd_uuid}</uuid><usage type='ceph'> <name>client.#{rbd_user} secret</name> </usage> </secret>"
+          secret_xml_file.run_action :create
```

Comment:
I guess I'm missing something, but why don't you simply do:
```
secret_file_path = "/etc/ceph/ceph-secret-#{rbd_uuid}.xml"
secret_file_content = "<secret ephemeral='no' private='no'> <uuid>#{rbd_uuid}</uuid><usage type='ceph'> <name>client.#{rbd_user} secret</name> </usage> </secret>"
File.write(secret_file_path, secret_file_content)
```

I know this won't be the same permission for the file, but there's no real secret in this file.